keps/sig-node/3759-kep-template/README.md

```diff
@@ -0,0 +1,1209 @@
+fork of Kubernetes with this sidecar functionality added (not all
+implementations are the same, but more than one company has a fork for this).
+
+### Problems: jobs with sidecar containers
+
+Imagine you have a Job with two containers: one which does the main processing of the job and the other is just a sidecar facilitating it. This sidecar could be a service mesh, a metrics gathering statsd server, etc.
+
+When the main processing finishes, the pod won't terminate until the sidecar container finishes too. This is problematic for sidecar containers that run continuously.
+
+There is no simple way to handle this on Kubernetes today. There are work-arounds for this problem, most of them consist of some form of coupling between the containers to add some logic where a container that finishes communicates it so other containers can react. But it gets tricky when you have more than one sidecar container or want to auto-inject sidecars.
+
+The sidecar will also not be restarted when it was OOM killed, which may render the pod unusable if the sidecar provided secure communication to other containers. 
+
+### Problems: logging/metrics sidecar
+
+A logging sidecar should start before several other containers, to not lose logs
+from the startup of other applications. Let's call "main" container the app that
+will log and "logging" container the sidecar that will facilitate it.
+
+If the logging container starts after the main app, some logs can be lost.
+Furthermore, if the logging container is not yet started and the main app
+crashes on startup, those logs can be lost (depends if logs go to a shared
+volume or over the network on localhost, etc.). While you can modify your
+application to handle this scenario during startup (as it is probably the change
+you need to do to handle sidecar crashes), for shutdown this approach won't
+work.
+
+On shutdown the ordering behavior is arguably more important: if the logging
+container is stopped first, logs for other containers are lost. No matter if
+those containers queue them and retry to send them to the logging container, or
+if they are persisted to a shared volume. The logging container is already
+killed and will not be restarted, as the pod is shutting down. In these cases,
+logs are lost.
+
+The same things regarding startup and shutdown apply for a metrics container.
+
+### Problems: service mesh
+
+Service mesh presents a similar problem: you want the service mesh container to
+be running and ready before other containers start, so that any inbound/outbound
+connections that a container can initiate goes through the service mesh.
+
+A similar problem happens for shutdown: if the service mesh container is
+terminated prior to the other containers, outgoing traffic from other apps will
+be blackholed or not use the service mesh.
+
+However, as none of these are possible to guarantee, most service meshes (like
+Linkerd and Istio), need to do several hacks to have the basic functionality.
```

Comment:
Not a strong opinion, but potentially enumerate some of the common hacks to show their downsides and how this KEP solves those problems?

e.g. One method that tries to solve this ordering problem is a "startup taint".  Here, the node is tainted on startup and a pod which tolerates the taint removes the taint when it's finished initializing to allow other pods to schedule. This isn't reliable either generally since other pods can also tolerate the taint and schedule before initializing is complete. It also doesn't help with the shutdown ordering case.

Reply:
this is about containers ordering inside the Pod, not Pods ordering on the Node.